### PR TITLE
string compare and return

### PR DESCRIPTION
**Pull request by ethancaballero**
- support for comparing strings and returning strings

### CODE REVIEW

File: myia/opt/__init__.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 """Optimization submodule."""
 
-from .clean import simplify_types, type_to_tag  # noqa
+from .clean import _strmap_tag, simplify_types, str_to_tag, type_to_tag  # noqa
```
Comment (abergeron) [inline myia/opt/__init__.py]: Don't expose private stuff.
```suggestion
from .clean import simplify_types, str_to_tag, type_to_tag  # noqa
```

File: myia/opt/clean.py [hunk 2]
```diff
@@ -35,11 +35,34 @@ def type_to_tag(t):
...
 @abstract_clone.variant
 def _reabs(self, a: AbstractClassBase):
     return (yield AbstractTuple)(self(x) for x in a.attributes.values())
 
 
+@overload  # noqa: F811
+def _reabs(self, a: AbstractScalar):
+    if a.values[TYPE] == String:
+        v = a.values[VALUE]
+        if isinstance(v, str):
```
Comment (abergeron) [inline myia/opt/clean.py]: This should always be true.  Make it an assert or remove it.

File: myia/pipeline/steps.py [hunk 3]
```diff
@@ -668,7 +678,11 @@ def convert_result(self, res, orig_t, vm_t: AbstractTuple, backend,
 @overload  # noqa: F811
 def convert_result(self, arg, orig_t, vm_t: AbstractScalar, backend,
                    return_backend):
-    return backend.to_scalar(arg)
+    ret = backend.to_scalar(arg)
+    if orig_t.values[TYPE] == dtype.String and \
+            vm_t.values[TYPE] == dtype.Int[64]:
```
Comment (abergeron) [inline myia/pipeline/steps.py]: You don't need to check that vm_t is Int[64].  This is implicit.

File: myia/pipeline/steps.py [hunk 4]
```diff
@@ -614,11 +618,17 @@ def convert_arg(self, arg, orig_t: AbstractScalar, backend):
...
+        if not isinstance(arg, str):
+            raise MyiaInputTypeError(f'Expected string')
+        arg = str_to_tag(arg)
+        t = dtype.Int[64]
     else:
         raise MyiaInputTypeError(f'Invalid type: {t}')
-    expected_value = orig_t.values[VALUE]
-    if expected_value is not ANYTHING and expected_value != arg:
-        raise MyiaInputTypeError(f'Invalid value: {arg}')
+    if not issubclass(orig_t.values[TYPE], dtype.String):
```
Comment (breuleux) [inline myia/pipeline/steps.py]: Why is that check here? If the program is compiled assuming that we are giving the string "hello" as a parameter, then we should check that this is the right string.

**Pull request merged**